### ISSUE TITLE
resolved: checking and using mtu size of socketcan interfaces

### ISSUE THREAD

**Issue by sknsean** [systemd/systemd#19396]
**systemd version the issue has been seen with**
 > `248`

**Used distribution**
 > `Yocto`

**Linux kernel version used** (`uname -a`)
 > `Linux iwg26-v3-ag 5.10.27`

**CPU architecture issue was seen on**
 > `arm`

**Expected behaviour you didn't see**
 > resolved sending UDP DNS requests

**Unexpected behaviour you saw**
 > resolved sending TCP DNS reguests

**Steps to reproduce the problem**
 > let interface's stay unmanaged by networkd and have a socketcan interface. Then resolved will send TCP DNS requests.

**Additional program output to the terminal or log subsystem illustrating the issue**
Debug log from resolved:
```
Emitting UDP, link MTU is 16, socket MTU is 0, minimal MTU is 40
Sending query via TCP since it is too large.
```

Resolved ends up in: `manager_find_mtu()`
I can mitigate the issue by adding:
```
+                if (!strncmp(l->ifname, "can", 3))
+                        continue;
+
```
But it's probably not the right place, to ignore the can interface :)

Links on target:
```
root@iwg26-v3-ag:/data/root# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: can0: <NOARP,UP,LOWER_UP,ECHO> mtu 16 qdisc pfifo_fast state UP mode DEFAULT group default qlen 10
    link/can 
3: can1: <NOARP,UP,LOWER_UP,ECHO> mtu 16 qdisc pfifo_fast state UP mode DEFAULT group default qlen 10
    link/can 
4: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 1a:28:d4:a1:85:b3 brd ff:ff:ff:ff:ff:ff
5: wwan0: <POINTOPOINT,MULTICAST,NOARP> mtu 1360 qdisc fq_codel state DOWN mode DEFAULT group default qlen 1000
    link/none
    ```

**Comment by poettering** [comment]
Well, but the the link MTU is 16 then it doesn't make sense to send UDP. So resolved is doing things correctly, no? Where's the problem?

**Comment by sknsean** [comment]
Well, I'm not trying to resolve DNS over socketcan :)
Default route is via the wwan0 interface...

**Comment by tlbdk** [comment]
I guess the issue here is that resolved does not know what interface the DNS packet will be sent on and has to fallback to guessing the interface by looking at the MTU size for all available interfaces and picking the lowest.

One could argue that picking the lowest might be a bit too pessimistic since it could just be a weird interfaces used something  that would never talk IP as in our embedded automotive use case. Maybe it would be possible to only consider interfaces that had an IP number or add an option to specify the possible interfaces the DNS packet might travel.

Alternatively resolved could lookup the route over netlink like what the "ip route get <address>" command does:
https://github.com/shemminger/iproute2/blob/9f366536edb5158343152604e82b968be46dbf26/ip/iproute.c#L2063

**Comment by poettering** [comment]
Fix posted in #19585. Any chance you can give it a try?
